nx/guides/modoc_agg_nx_01.livemd

```diff
@@ -0,0 +1,826 @@
+
+<!-- livebook:{"break_markdown":true} -->
+
+###### Aggregate along the first axis `axis: :x` (equivalently `axis: 0`).
+
+The shape of our original tensor  `t`  is `[x: 2, y: 3, z: 2, t: 3]`. When we aggregate along the axis `:x`, it collapses and the shape of the resultant tensor `r` is `[y: 3,z: 2,t: 3]`. The rule is that every axis before the selected one will remain in the structure of the tensor as well as every axis after the selected one.
+
+How? You collect from each slice of `x` the elements that have the _same_ remaining indexes "on the right" $y,z,t$. In the case of our tensor, you have 2 `x`-slices so you build a sublist of 2 elements, and then apply the aggregating function on it. This gives you the value of the resulting tensor $r$ at the location $r[y][z][t]$:
+
+$$
+\rm{agg}\big(t[0][y][z][t], t[1][y][z][t]\big) = r[y][z][t]
+$$
+
+The aggregating function used below is `argmin`: the index of the smallest element of the  sublist. In case of several occurences, the `:tie_break` attribute takes the lowest index.
+
+$$
+\begin{bmatrix} x=0
+  \begin{bmatrix} y=0
+
+```
+\begin{bmatrix}
+  & \boxed{1}, 2, 3, & \\
+  & 1, -2, 3, &
+\end{bmatrix}
+\\ 
+y=1
+\begin{bmatrix} 
+  & 4, -3, 2 & \\
+  & -4, 3, 2 &
+\end{bmatrix}
+\\ 
+y=2
+\begin{bmatrix}
+  & 5, -1, 3 &\\
+  & -5, 1, 3 &
+\end{bmatrix}
+```
+
+\end{bmatrix}
+  \\
+  x=1
+  \begin{bmatrix} y=0
+
+```
+\begin{bmatrix}
+  & \boxed{4}, 6, 1 & \\
+  & 4, -6, 1 &
+\end{bmatrix}
+\\ 
+y=1
+\begin{bmatrix}
+  & 1, 2, 3 & \\
+  & 1, -2, 3 &
+\end{bmatrix}
+\\ 
+y=2
+\begin{bmatrix}
+  & 4, -3, 2 & \\
+  & -4, 3, 2 &
+\end{bmatrix}
+```
+
+\end{bmatrix}
+\end{bmatrix}
+
+\to
+  \begin{bmatrix} y=0
+
+```
+\begin{bmatrix}
+  & \boxed{\rm{agg}(1,4)}, \rm{agg}(2,6), \rm{agg}(3, 1)& \\
+  & \rm{agg}(1,4), \rm{agg}(-2,-6), \rm{agg}(3,1) &
+\end{bmatrix}
+\\ 
+y=1
+\begin{bmatrix} 
+  & \rm{agg}(4,1), \rm{agg}(-3,2), \rm{agg}(2,3) & \\
+  & \rm{agg}(-4,1), \rm{agg}(3,-2), \rm{agg}(2,3) &
+\end{bmatrix}
+\\ 
+y=2
+\begin{bmatrix}
+  & \rm{agg}(5,4), \rm{agg}(-1,-3), \rm{agg}(3,2) &\\
+  & \rm{agg}(-5,-4), \rm{agg}(1,3), \rm{agg}(3,2) &
+\end{bmatrix}
+```
+
+\end{bmatrix}
+
+\xrightarrow[]{\rm{argmin}}
+
+\begin{bmatrix}
+
+```
+\begin{bmatrix}
+  & 0,0, 1 & \\
+  & 0, 1,1 &
+\end{bmatrix}
+\\ 
+\begin{bmatrix} 
+  & 1, 0,0 & \\
+  & 0,1,0 &
+\end{bmatrix}
+\\ 
+\begin{bmatrix}
+  & 1, 1,1 &\\
+  & 0, 0,1 &
+\end{bmatrix}
+```
+
+\end{bmatrix}
+$$
+
+```elixir
+Nx.argmin(t, axis: :x)
+```
+
+###### Aggregate along the second axis `axis: :y` (equivalently `axis: 1`) .
+
+The axis `y` will collapse and the shape of the resultant tensor `r` is `[x: 2, z: 2, t: 3]`.
+
+How? The axis $x$ is before the selected one $y$ so it will remain. We therefor apply the procedure as above for _each_ sub $x$-slice.
```

Comment:
```suggestion
How? The axis $x$ is before the selected one $y$ so it will remain. We therefore apply the procedure as above for _each_ sub $x$-slice.
```

It's not actually for each x slice though. You're actually travelling each arrangement of {x, z, t} possible values. For each of those you take one full slice over `y`. Showing equivalent elixir code might be more clear to the reader:

```elixir
# The code below is conceptually equivalent to how aggregation subsets are taken:
for x <- 0..(Nx.axis_size(tensor, :x) - 1), z <- 0..(Nx.axis_size(tensor, :z) - 1), t <- 0..(Nx.axis_size(tensor, :t) - 1) do
  tensor[[x, 0..-1//1, z, t]]
end
```

And actually showing the result from a code cell might also help. Example:

```elixir
iex(13)> t = Nx.iota({2, 4, 3})
#Nx.Tensor<
  s64[2][4][3]
  [
    [
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [9, 10, 11]
    ],
    [
      [12, 13, 14],
      [15, 16, 17],
      [18, 19, 20],
      [21, 22, 23]
    ]
  ]
>
iex(15)> Nx.sum(t, axes: [1])
#Nx.Tensor<
  s64[2][3]
  [
    [18, 22, 26],
    [66, 70, 74]
  ]
>
iex(16)> for x <- 0..(Nx.axis_size(t, 0) - 1), z <- 0..(Nx.axis_size(t, 2) - 1) do
...(16)> Nx.sum(t[[x, 0..-1//1, z]])
...(16)> end
[
  #Nx.Tensor<
    s64
    18
  >,
  #Nx.Tensor<
    s64
    22
  >,
  #Nx.Tensor<
    s64
    26
  >,
  #Nx.Tensor<
    s64
    66
  >,
  #Nx.Tensor<
    s64
    70
  >,
  #Nx.Tensor<
    s64
    74
  >
]
```